// @buf/evmos_evmos.bufbuild_es/evmos/erc20/v1/query_pb.js
// Copyright Tharsis Labs Ltd.(Evmos)
// SPDX-License-Identifier:ENCL-1.0(https://github.com/evmos/evmos/blob/main/LICENSE)

// @generated by protoc-gen-es v1.2.0
// @generated from file evmos/erc20/v1/query.proto (package evmos.erc20.v1, syntax proto3)
/* eslint-disable */
// @ts-nocheck

import { proto3 } from "@bufbuild/protobuf";
import { PageRequest, PageResponse } from "@buf/cosmos_cosmos-sdk.bufbuild_es/cosmos/base/query/v1beta1/pagination_pb.js";
import { TokenPair } from "./erc20_pb.js";
import { Params } from "./genesis_pb.js";

/**
 * QueryTokenPairsRequest is the request type for the Query/TokenPairs RPC
 * method.
 *
 * @generated from message evmos.erc20.v1.QueryTokenPairsRequest
 */
export const QueryTokenPairsRequest = proto3.makeMessageType(
  "evmos.erc20.v1.QueryTokenPairsRequest",
  () => [
    { no: 1, name: "pagination", kind: "message", T: PageRequest },
  ],
);

/**
 * QueryTokenPairsResponse is the response type for the Query/TokenPairs RPC
 * method.
 *
 * @generated from message evmos.erc20.v1.QueryTokenPairsResponse
 */
export const QueryTokenPairsResponse = proto3.makeMessageType(
  "evmos.erc20.v1.QueryTokenPairsResponse",
  () => [
    { no: 1, name: "token_pairs", kind: "message", T: TokenPair, repeated: true },
    { no: 2, name: "pagination", kind: "message", T: PageResponse },
  ],
);

/**
 * QueryTokenPairRequest is the request type for the Query/TokenPair RPC method.
 *
 * @generated from message evmos.erc20.v1.QueryTokenPairRequest
 */
export const QueryTokenPairRequest = proto3.makeMessageType(
  "evmos.erc20.v1.QueryTokenPairRequest",
  () => [
    { no: 1, name: "token", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ],
);

/**
 * QueryTokenPairResponse is the response type for the Query/TokenPair RPC
 * method.
 *
 * @generated from message evmos.erc20.v1.QueryTokenPairResponse
 */
export const QueryTokenPairResponse = proto3.makeMessageType(
  "evmos.erc20.v1.QueryTokenPairResponse",
  () => [
    { no: 1, name: "token_pair", kind: "message", T: TokenPair },
  ],
);

/**
 * QueryParamsRequest is the request type for the Query/Params RPC method.
 *
 * @generated from message evmos.erc20.v1.QueryParamsRequest
 */
export const QueryParamsRequest = proto3.makeMessageType(
  "evmos.erc20.v1.QueryParamsRequest",
  [],
);

/**
 * QueryParamsResponse is the response type for the Query/Params RPC
 * method.
 *
 * @generated from message evmos.erc20.v1.QueryParamsResponse
 */
export const QueryParamsResponse = proto3.makeMessageType(
  "evmos.erc20.v1.QueryParamsResponse",
  () => [
    { no: 1, name: "params", kind: "message", T: Params },
  ],
);
