// @buf/cosmos_cosmos-sdk.bufbuild_es/cosmos/gov/v1beta1/tx_pb.js
// @generated by protoc-gen-es v1.2.0
// @generated from file cosmos/gov/v1beta1/tx.proto (package cosmos.gov.v1beta1, syntax proto3)
/* eslint-disable */
// @ts-nocheck

import { Any, proto3 } from "@bufbuild/protobuf";
import { Coin } from "../../base/v1beta1/coin_pb.js";
import { VoteOption, WeightedVoteOption } from "./gov_pb.js";

/**
 * MsgSubmitProposal defines an sdk.Msg type that supports submitting arbitrary
 * proposal Content.
 *
 * @generated from message cosmos.gov.v1beta1.MsgSubmitProposal
 */
export const MsgSubmitProposal = proto3.makeMessageType(
  "cosmos.gov.v1beta1.MsgSubmitProposal",
  () => [
    { no: 1, name: "content", kind: "message", T: Any },
    { no: 2, name: "initial_deposit", kind: "message", T: Coin, repeated: true },
    { no: 3, name: "proposer", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ],
);

/**
 * MsgSubmitProposalResponse defines the Msg/SubmitProposal response type.
 *
 * @generated from message cosmos.gov.v1beta1.MsgSubmitProposalResponse
 */
export const MsgSubmitProposalResponse = proto3.makeMessageType(
  "cosmos.gov.v1beta1.MsgSubmitProposalResponse",
  () => [
    { no: 1, name: "proposal_id", kind: "scalar", T: 4 /* ScalarType.UINT64 */ },
  ],
);

/**
 * MsgVote defines a message to cast a vote.
 *
 * @generated from message cosmos.gov.v1beta1.MsgVote
 */
export const MsgVote = proto3.makeMessageType(
  "cosmos.gov.v1beta1.MsgVote",
  () => [
    { no: 1, name: "proposal_id", kind: "scalar", T: 4 /* ScalarType.UINT64 */ },
    { no: 2, name: "voter", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "option", kind: "enum", T: proto3.getEnumType(VoteOption) },
  ],
);

/**
 * MsgVoteResponse defines the Msg/Vote response type.
 *
 * @generated from message cosmos.gov.v1beta1.MsgVoteResponse
 */
export const MsgVoteResponse = proto3.makeMessageType(
  "cosmos.gov.v1beta1.MsgVoteResponse",
  [],
);

/**
 * MsgVoteWeighted defines a message to cast a vote.
 *
 * Since: cosmos-sdk 0.43
 *
 * @generated from message cosmos.gov.v1beta1.MsgVoteWeighted
 */
export const MsgVoteWeighted = proto3.makeMessageType(
  "cosmos.gov.v1beta1.MsgVoteWeighted",
  () => [
    { no: 1, name: "proposal_id", kind: "scalar", T: 4 /* ScalarType.UINT64 */ },
    { no: 2, name: "voter", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "options", kind: "message", T: WeightedVoteOption, repeated: true },
  ],
);

/**
 * MsgVoteWeightedResponse defines the Msg/VoteWeighted response type.
 *
 * Since: cosmos-sdk 0.43
 *
 * @generated from message cosmos.gov.v1beta1.MsgVoteWeightedResponse
 */
export const MsgVoteWeightedResponse = proto3.makeMessageType(
  "cosmos.gov.v1beta1.MsgVoteWeightedResponse",
  [],
);

/**
 * MsgDeposit defines a message to submit a deposit to an existing proposal.
 *
 * @generated from message cosmos.gov.v1beta1.MsgDeposit
 */
export const MsgDeposit = proto3.makeMessageType(
  "cosmos.gov.v1beta1.MsgDeposit",
  () => [
    { no: 1, name: "proposal_id", kind: "scalar", T: 4 /* ScalarType.UINT64 */ },
    { no: 2, name: "depositor", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "amount", kind: "message", T: Coin, repeated: true },
  ],
);

/**
 * MsgDepositResponse defines the Msg/Deposit response type.
 *
 * @generated from message cosmos.gov.v1beta1.MsgDepositResponse
 */
export const MsgDepositResponse = proto3.makeMessageType(
  "cosmos.gov.v1beta1.MsgDepositResponse",
  [],
);
