// @buf/cosmos_cosmos-sdk.bufbuild_es/cosmos/circuit/v1/tx_pb.js
// @generated by protoc-gen-es v1.0.0 with parameter "rewrite_imports=./google/api/**/*_pb.js:@buf/googleapis_googleapis.bufbuild_es,rewrite_imports=./gogoproto/**/*_pb.js:@buf/cosmos_gogo-proto.bufbuild_es,rewrite_imports=./cosmos_proto/**/*_pb.js:@buf/cosmos_cosmos-proto.bufbuild_es"
// @generated from file cosmos/circuit/v1/tx.proto (package cosmos.circuit.v1, syntax proto3)
/* eslint-disable */
// @ts-nocheck

import { proto3 } from "@bufbuild/protobuf";
import { Permissions } from "./types_pb.js";

/**
 * MsgAuthorizeCircuitBreaker defines the Msg/AuthorizeCircuitBreaker request type.
 *
 * @generated from message cosmos.circuit.v1.MsgAuthorizeCircuitBreaker
 */
export const MsgAuthorizeCircuitBreaker = proto3.makeMessageType(
  "cosmos.circuit.v1.MsgAuthorizeCircuitBreaker",
  () => [
    { no: 1, name: "granter", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "grantee", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "permissions", kind: "message", T: Permissions },
  ],
);

/**
 * MsgAuthorizeCircuitBreakerResponse defines the Msg/AuthorizeCircuitBreaker response type.
 *
 * @generated from message cosmos.circuit.v1.MsgAuthorizeCircuitBreakerResponse
 */
export const MsgAuthorizeCircuitBreakerResponse = proto3.makeMessageType(
  "cosmos.circuit.v1.MsgAuthorizeCircuitBreakerResponse",
  () => [
    { no: 1, name: "success", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
  ],
);

/**
 * MsgTripCircuitBreaker defines the Msg/TripCircuitBreaker request type.
 *
 * @generated from message cosmos.circuit.v1.MsgTripCircuitBreaker
 */
export const MsgTripCircuitBreaker = proto3.makeMessageType(
  "cosmos.circuit.v1.MsgTripCircuitBreaker",
  () => [
    { no: 1, name: "authority", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "msg_type_urls", kind: "scalar", T: 9 /* ScalarType.STRING */, repeated: true },
  ],
);

/**
 * MsgTripCircuitBreakerResponse defines the Msg/TripCircuitBreaker response type.
 *
 * @generated from message cosmos.circuit.v1.MsgTripCircuitBreakerResponse
 */
export const MsgTripCircuitBreakerResponse = proto3.makeMessageType(
  "cosmos.circuit.v1.MsgTripCircuitBreakerResponse",
  () => [
    { no: 1, name: "success", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
  ],
);

/**
 * MsgResetCircuitBreaker defines the Msg/ResetCircuitBreaker request type.
 *
 * @generated from message cosmos.circuit.v1.MsgResetCircuitBreaker
 */
export const MsgResetCircuitBreaker = proto3.makeMessageType(
  "cosmos.circuit.v1.MsgResetCircuitBreaker",
  () => [
    { no: 1, name: "authority", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "msg_type_urls", kind: "scalar", T: 9 /* ScalarType.STRING */, repeated: true },
  ],
);

/**
 * MsgResetCircuitBreakerResponse defines the Msg/ResetCircuitBreaker response type.
 *
 * @generated from message cosmos.circuit.v1.MsgResetCircuitBreakerResponse
 */
export const MsgResetCircuitBreakerResponse = proto3.makeMessageType(
  "cosmos.circuit.v1.MsgResetCircuitBreakerResponse",
  () => [
    { no: 1, name: "success", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
  ],
);
