// @buf/cosmos_cosmos-sdk.bufbuild_es/cosmos/authz/v1beta1/query_pb.js
// Since: cosmos-sdk 0.43

// @generated by protoc-gen-es v1.2.0
// @generated from file cosmos/authz/v1beta1/query.proto (package cosmos.authz.v1beta1, syntax proto3)
/* eslint-disable */
// @ts-nocheck

import { proto3 } from "@bufbuild/protobuf";
import { PageRequest, PageResponse } from "../../base/query/v1beta1/pagination_pb.js";
import { Grant, GrantAuthorization } from "./authz_pb.js";

/**
 * QueryGrantsRequest is the request type for the Query/Grants RPC method.
 *
 * @generated from message cosmos.authz.v1beta1.QueryGrantsRequest
 */
export const QueryGrantsRequest = proto3.makeMessageType(
  "cosmos.authz.v1beta1.QueryGrantsRequest",
  () => [
    { no: 1, name: "granter", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "grantee", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "msg_type_url", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 4, name: "pagination", kind: "message", T: PageRequest },
  ],
);

/**
 * QueryGrantsResponse is the response type for the Query/Authorizations RPC method.
 *
 * @generated from message cosmos.authz.v1beta1.QueryGrantsResponse
 */
export const QueryGrantsResponse = proto3.makeMessageType(
  "cosmos.authz.v1beta1.QueryGrantsResponse",
  () => [
    { no: 1, name: "grants", kind: "message", T: Grant, repeated: true },
    { no: 2, name: "pagination", kind: "message", T: PageResponse },
  ],
);

/**
 * QueryGranterGrantsRequest is the request type for the Query/GranterGrants RPC method.
 *
 * @generated from message cosmos.authz.v1beta1.QueryGranterGrantsRequest
 */
export const QueryGranterGrantsRequest = proto3.makeMessageType(
  "cosmos.authz.v1beta1.QueryGranterGrantsRequest",
  () => [
    { no: 1, name: "granter", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "pagination", kind: "message", T: PageRequest },
  ],
);

/**
 * QueryGranterGrantsResponse is the response type for the Query/GranterGrants RPC method.
 *
 * @generated from message cosmos.authz.v1beta1.QueryGranterGrantsResponse
 */
export const QueryGranterGrantsResponse = proto3.makeMessageType(
  "cosmos.authz.v1beta1.QueryGranterGrantsResponse",
  () => [
    { no: 1, name: "grants", kind: "message", T: GrantAuthorization, repeated: true },
    { no: 2, name: "pagination", kind: "message", T: PageResponse },
  ],
);

/**
 * QueryGranteeGrantsRequest is the request type for the Query/IssuedGrants RPC method.
 *
 * @generated from message cosmos.authz.v1beta1.QueryGranteeGrantsRequest
 */
export const QueryGranteeGrantsRequest = proto3.makeMessageType(
  "cosmos.authz.v1beta1.QueryGranteeGrantsRequest",
  () => [
    { no: 1, name: "grantee", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "pagination", kind: "message", T: PageRequest },
  ],
);

/**
 * QueryGranteeGrantsResponse is the response type for the Query/GranteeGrants RPC method.
 *
 * @generated from message cosmos.authz.v1beta1.QueryGranteeGrantsResponse
 */
export const QueryGranteeGrantsResponse = proto3.makeMessageType(
  "cosmos.authz.v1beta1.QueryGranteeGrantsResponse",
  () => [
    { no: 1, name: "grants", kind: "message", T: GrantAuthorization, repeated: true },
    { no: 2, name: "pagination", kind: "message", T: PageResponse },
  ],
);
