// @tharsis/eip712/dist/messages/staking.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMsgWithdrawValidatorCommission = exports.MSG_WITHDRAW_VALIDATOR_COMMISSION_TYPES = exports.createMsgWithdrawDelegatorReward = exports.MSG_WITHDRAW_DELEGATOR_REWARD_TYPES = exports.createMsgUndelegate = exports.MSG_UNDELEGATE_TYPES = exports.createMsgBeginRedelegate = exports.MSG_BEGIN_REDELEGATE_TYPES = exports.createMsgDelegate = exports.MSG_DELEGATE_TYPES = void 0;
exports.MSG_DELEGATE_TYPES = {
    MsgValue: [
        { name: 'delegator_address', type: 'string' },
        { name: 'validator_address', type: 'string' },
        { name: 'amount', type: 'TypeAmount' },
    ],
    TypeAmount: [
        { name: 'denom', type: 'string' },
        { name: 'amount', type: 'string' },
    ],
};
function createMsgDelegate(delegatorAddress, validatorAddress, amount, denom) {
    return {
        type: 'cosmos-sdk/MsgDelegate',
        value: {
            amount: {
                amount,
                denom,
            },
            delegator_address: delegatorAddress,
            validator_address: validatorAddress,
        },
    };
}
exports.createMsgDelegate = createMsgDelegate;
exports.MSG_BEGIN_REDELEGATE_TYPES = {
    MsgValue: [
        { name: 'delegator_address', type: 'string' },
        { name: 'validator_src_address', type: 'string' },
        { name: 'validator_dst_address', type: 'string' },
        { name: 'amount', type: 'TypeAmount' },
    ],
    TypeAmount: [
        { name: 'denom', type: 'string' },
        { name: 'amount', type: 'string' },
    ],
};
function createMsgBeginRedelegate(delegatorAddress, validatorSrcAddress, validatorDstAddress, amount, denom) {
    return {
        type: 'cosmos-sdk/MsgBeginRedelegate',
        value: {
            amount: {
                amount,
                denom,
            },
            delegator_address: delegatorAddress,
            validator_src_address: validatorSrcAddress,
            validator_dst_address: validatorDstAddress,
        },
    };
}
exports.createMsgBeginRedelegate = createMsgBeginRedelegate;
exports.MSG_UNDELEGATE_TYPES = {
    MsgValue: [
        { name: 'delegator_address', type: 'string' },
        { name: 'validator_address', type: 'string' },
        { name: 'amount', type: 'TypeAmount' },
    ],
    TypeAmount: [
        { name: 'denom', type: 'string' },
        { name: 'amount', type: 'string' },
    ],
};
function createMsgUndelegate(delegatorAddress, validatorAddress, amount, denom) {
    return {
        type: 'cosmos-sdk/MsgUndelegate',
        value: {
            amount: {
                amount,
                denom,
            },
            delegator_address: delegatorAddress,
            validator_address: validatorAddress,
        },
    };
}
exports.createMsgUndelegate = createMsgUndelegate;
exports.MSG_WITHDRAW_DELEGATOR_REWARD_TYPES = {
    MsgValue: [
        { name: 'delegator_address', type: 'string' },
        { name: 'validator_address', type: 'string' },
    ],
};
function createMsgWithdrawDelegatorReward(delegatorAddress, validatorAddress) {
    return {
        type: 'cosmos-sdk/MsgWithdrawDelegationReward',
        value: {
            delegator_address: delegatorAddress,
            validator_address: validatorAddress,
        },
    };
}
exports.createMsgWithdrawDelegatorReward = createMsgWithdrawDelegatorReward;
exports.MSG_WITHDRAW_VALIDATOR_COMMISSION_TYPES = {
    MsgValue: [{ name: 'validator_address', type: 'string' }],
};
function createMsgWithdrawValidatorCommission(validatorAddress) {
    return {
        type: 'cosmos-sdk/MsgWithdrawValidatorCommission',
        value: {
            validator_address: validatorAddress,
        },
    };
}
exports.createMsgWithdrawValidatorCommission = createMsgWithdrawValidatorCommission;
