// @buf/evmos_evmos.bufbuild_es/evmos/revenue/v1/tx_pb.js
// Copyright Tharsis Labs Ltd.(Evmos)
// SPDX-License-Identifier:LGPL-3.0-only

// @generated by protoc-gen-es v1.2.0
// @generated from file evmos/revenue/v1/tx.proto (package evmos.revenue.v1, syntax proto3)
/* eslint-disable */
// @ts-nocheck

import { proto3 } from "@bufbuild/protobuf";
import { Params } from "./genesis_pb.js";

/**
 * MsgRegisterRevenue defines a message that registers a Revenue
 *
 * @generated from message evmos.revenue.v1.MsgRegisterRevenue
 */
export const MsgRegisterRevenue = proto3.makeMessageType(
  "evmos.revenue.v1.MsgRegisterRevenue",
  () => [
    { no: 1, name: "contract_address", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "deployer_address", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "withdrawer_address", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 4, name: "nonces", kind: "scalar", T: 4 /* ScalarType.UINT64 */, repeated: true },
  ],
);

/**
 * MsgRegisterRevenueResponse defines the MsgRegisterRevenue response type
 *
 * @generated from message evmos.revenue.v1.MsgRegisterRevenueResponse
 */
export const MsgRegisterRevenueResponse = proto3.makeMessageType(
  "evmos.revenue.v1.MsgRegisterRevenueResponse",
  [],
);

/**
 * MsgUpdateRevenue defines a message that updates the withdrawer address for a
 * registered Revenue
 *
 * @generated from message evmos.revenue.v1.MsgUpdateRevenue
 */
export const MsgUpdateRevenue = proto3.makeMessageType(
  "evmos.revenue.v1.MsgUpdateRevenue",
  () => [
    { no: 1, name: "contract_address", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "deployer_address", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "withdrawer_address", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ],
);

/**
 * MsgUpdateRevenueResponse defines the MsgUpdateRevenue response type
 *
 * @generated from message evmos.revenue.v1.MsgUpdateRevenueResponse
 */
export const MsgUpdateRevenueResponse = proto3.makeMessageType(
  "evmos.revenue.v1.MsgUpdateRevenueResponse",
  [],
);

/**
 * MsgCancelRevenue defines a message that cancels a registered Revenue
 *
 * @generated from message evmos.revenue.v1.MsgCancelRevenue
 */
export const MsgCancelRevenue = proto3.makeMessageType(
  "evmos.revenue.v1.MsgCancelRevenue",
  () => [
    { no: 1, name: "contract_address", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "deployer_address", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ],
);

/**
 * MsgCancelRevenueResponse defines the MsgCancelRevenue response type
 *
 * @generated from message evmos.revenue.v1.MsgCancelRevenueResponse
 */
export const MsgCancelRevenueResponse = proto3.makeMessageType(
  "evmos.revenue.v1.MsgCancelRevenueResponse",
  [],
);

/**
 * MsgUpdateParams defines a Msg for updating the x/revenue module parameters.
 *
 * @generated from message evmos.revenue.v1.MsgUpdateParams
 */
export const MsgUpdateParams = proto3.makeMessageType(
  "evmos.revenue.v1.MsgUpdateParams",
  () => [
    { no: 1, name: "authority", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "params", kind: "message", T: Params },
  ],
);

/**
 * MsgUpdateParamsResponse defines the response structure for executing a
 * MsgUpdateParams message.
 *
 * @generated from message evmos.revenue.v1.MsgUpdateParamsResponse
 */
export const MsgUpdateParamsResponse = proto3.makeMessageType(
  "evmos.revenue.v1.MsgUpdateParamsResponse",
  [],
);
